File: backend/domain/jobs.py
from abc import ABC, abstractmethod
import uuid
from uuid import UUID
from datetime import datetime
from dataclasses import dataclass
from typing import List, Optional, Tuple

@dataclass
class Address:
    country: str
    city: str
    state: str
    location: str

@dataclass
class Job:
    id: UUID
    company: str
    job_title: str
    address: Address
    last_updated: datetime
    apply_url: Optional[str] = None
    source_id: Optional[str] = None
    
    @classmethod
    def new(cls, company: str, job_title: str,
        address: Address, apply_url: str, source_id: str,
        last_updated: datetime = None) -> "Job":

        return cls(
            id = uuid.uuid4(),
            company = company,
            job_title = job_title,
            address = address,
            last_updated = last_updated or datetime.now(),
            apply_url = apply_url,
            source_id = source_id
        )
        
    def update(self, company: str, job_title: str, address: Address, apply_url: str,
               last_updated: datetime = None) -> None:
        self.company = company
        self.job_title = job_title
        self.address = address
        self.apply_url = apply_url
        self.last_updated = last_updated or datetime.now()

class JobRepository(ABC):
    @abstractmethod
    def save(self, job: Job) -> None:
        pass
    
    @abstractmethod
    def get_job_by_id(self, job_id: UUID) -> Optional[Job]:
        pass
    
    @abstractmethod
    def list_jobs(self, limit, last_key_value: Optional[str] = None) -> Tuple[List[Job], Optional[str]]:
        pass
    
    @abstractmethod
    def delete_job_by_id(self, job_id: UUID) -> bool:
        pass
    
    @abstractmethod
    def get_by_source_id(self, source_id: str) -> Optional[Job]:
        pass

File: webapp/src/hooks/jobs.ts
import useSWR from "swr";
import Config from "../config.ts";
import {BaseEntity, ClientResponse, fetcher, PaginatedResponse} from "./shared.ts";
import useSWRInfinite from "swr/infinite";

export interface Job extends BaseEntity {
    company: string;
    content?: string;
    job_title: string;
    token_count_llama: number;
    address: {
        country: string;
        city: string;
        state: string;
        location: string;
    },
    last_updated: string;
    apply_url: string;
    source_id: string;
}

export const useJobs = () => {
    const getKey = (pageIndex: number, previousPageData: PaginatedResponse<Job>) => {
        if (pageIndex === 0) return `${Config.API_ENDPOINT}/jobs?limit=10`;
        if (!previousPageData.last_key_value || previousPageData.items.length === 0) return null;
        return `${Config.API_ENDPOINT}/jobs?limit=10&last_key_value=${encodeURIComponent(previousPageData.last_key_value)}`;
    }

    const {
        data,
        error,
        isLoading,
        size,
        setSize
    } = useSWRInfinite<PaginatedResponse<Job>>(getKey, fetcher, { keepPreviousData: true });

    const flatData = data?.reduce<Job[]>((accum, current) => {
        return [...accum, ...current.items];
    }, [])

    return { data: flatData, error, isLoading, size, setSize }
}

export const useJob = (jobId?: string): ClientResponse<Job> => {
    const {
        data: jobData,
        error: jobError,
        isLoading: jobIsLoading
    } = useSWR<Job>(jobId ? `${Config.API_ENDPOINT}/jobs/${jobId}` : null, fetcher, {
        keepPreviousData: true
    });

    const {
        data: contentData,
        error: contentError,
        isLoading: contentIsLoading
    } = useSWR<string>(jobId ? `${Config.API_ENDPOINT}/jobs/${jobId}/description` : null, (url) => fetcher(url, {
        text: true
    }), {
        keepPreviousData: true
    })

    const data = jobData ? { ...jobData, content: contentData } : undefined;

    return {
        data,
        error: jobError || contentError,
        isLoading: contentIsLoading || jobIsLoading
    }
}